test/e2e/mattermost_test.go

```diff
@@ -90,9 +94,94 @@ func mattermostScaleTest(t *testing.T, f *framework.Framework, ctx *framework.Te
 
 	// wait for test-mm to reach 3 replicas
 	// return e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "test-mm", 3, retryInterval, timeout)
+
+	err = f.Client.Delete(goctx.TODO(), exampleMattermost)
+	assert.Nil(t, err)
+
 	return nil
 }
 
+func mattermostUpgradeTest(t *testing.T, f *framework.Framework, ctx *framework.TestCtx) error {
+	namespace, err := ctx.GetNamespace()
+	assert.Nil(t, err)
+
+	// create memcached custom resource
+	exampleMattermost := &operator.ClusterInstallation{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterInstallation",
+			APIVersion: "mattermost.com/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-mm2",
+			Namespace: namespace,
+		},
+		Spec: operator.ClusterInstallationSpec{
+			Image:            "mattermost/mattermost-enterprise-edition",
+			Version:          "5.10.0",
+			IngressName:      "test-example2.mattermost.dev",
+			Replicas:         1,
+			MinioStorageSize: "1Gi",
+		},
+	}
+
+	// use TestCtx's create helper to create the object and add a cleanup function for the new object
+	err = f.Client.Create(goctx.TODO(), exampleMattermost, &framework.CleanupOptions{TestContext: ctx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+	assert.Nil(t, err)
+
+	// wait for test-mm2 to reach 1 replicas
+	err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "test-mm2", 1, retryInterval, timeout)
+	assert.Nil(t, err)
+
+	err = f.Client.Get(goctx.TODO(), types.NamespacedName{Name: "test-mm2", Namespace: namespace}, exampleMattermost)
+	assert.Nil(t, err)
+
+	// Get the current pod
+	pods := corev1.PodList{}
+	opts := client.ListOptions{Namespace: namespace}
+	opts.SetLabelSelector("app=mattermost")
+	err = f.Client.List(context.TODO(), &opts, &pods)
+	assert.Nil(t, err)
+	assert.Equal(t, 1, len(pods.Items))
+
+	mmOldPod := &corev1.Pod{}
+	err = f.Client.Get(goctx.TODO(), types.NamespacedName{Name: pods.Items[0].Name, Namespace: namespace}, mmOldPod)
+	assert.Nil(t, err)
+
+	// Apply the new version
+	exampleMattermost.Spec.Version = "5.11.0"
+	err = f.Client.Update(goctx.TODO(), exampleMattermost)
+	assert.Nil(t, err)
+
+	// Wait for this pod be terminated
+	err = e2eutil.WaitForDeletion(t, f.Client.Client, mmOldPod, retryInterval, timeout)
+	assert.Nil(t, err)
+
+	// wait for deployment be completed
+	err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "test-mm2", 1, retryInterval, timeout)
+	assert.Nil(t, err)
+
+	time.Sleep(time.Second * 5)
```

Comment:
Do we need this sleep? The very next function seems to wait anyway.

Reply:
maybe not, removed